### PR TITLE
refactor: TypeId instead of string-based ResourceId

### PR DESCRIPTION
**Pull request by joonazan**
`std::any::TypeId` solves the same problem as the current string identifiers, while using less memory. It eliminates errors that can occur due to copy-paste or similar naming.

I'm not super happy with how building the name works but it is good enough in my opinion. I tried to add `const NAME: &'static str` to the interface but learned that constants can't depend on generic parameters of the trait impl. At least if the compiler is even a little bit smart it will concatenate the strings at compile time.

### CODE REVIEW

**Comment by joonazan** [comment]
How do I make spellcheck happy while still linking to the documentation of a type?

File: core/node/node_framework/src/wiring_layer.rs [hunk 1]
```diff
@@ -30,9 +30,9 @@ impl fmt::Debug for dyn WiringLayer {
 #[non_exhaustive]
 pub enum WiringError {
     #[error("Layer attempted to add resource {0}, but it is already provided")]
-    ResourceAlreadyProvided(ResourceId),
+    ResourceAlreadyProvided(String),
     #[error("Resource {0} is not provided")]
-    ResourceLacking(ResourceId),
+    ResourceLacking(String),
```
Comment (slowli) [inline core/node/node_framework/src/wiring_layer.rs:35]: If we treat errors are something that node framework users would actually use (i.e., having utility besides converting an error to a string), having `String` variants (i.e., only covering human-readability) is suboptimal. May make sense to add `ResourceId` as well.

**Comment by popzxc** [comment]
@joonazan to make spellchecker happy, you may simply put it to backquotes (e.g. \`TypeId\`) to indicate that it's not a word.

File: core/node/node_framework/src/service/context.rs [hunk 2]
```diff
@@ -108,29 +108,28 @@ impl<'a> ServiceContext<'a> {
...
                 })
                 .clone()
         };
 
-        let name = T::resource_id();
         // Check whether the resource is already available.
-        if let Some(resource) = self.service.resources.get(&name) {
-            tracing::info!("Layer {} has requested resource {}", self.layer, name);
+        if let Some(resource) = self.service.resources.get(&ResourceId::of::<T>()) {
+            tracing::info!("Layer {} has requested resource {}", self.layer, T::name());
```
[the remark this answers is not in the document]
Comment (joonazan) [reply core/node/node_framework/src/service/context.rs:120]: I thought of the same thing. Your proposed solution is very good.

File: core/node/node_framework/src/resource/lazy_resource.rs [hunk 3]
```diff
@@ -18,8 +18,8 @@ pub struct LazyResource<T: Resource> {
 }
 
 impl<T: Resource> Resource for LazyResource<T> {
-    fn resource_id() -> ResourceId {
-        ResourceId::new("lazy") + T::resource_id()
+    fn name() -> String {
+        format!("lazy {}", T::name())
```
[the remark this answers is not in the document]
Comment (joonazan) [reply core/node/node_framework/src/resource/lazy_resource.rs:22]: I think String is a good type for info that is only meant for human consumption.

**Review by popzxc — DISMISSED** [review]
There may be some things that may be polished, but I think we can work on it separately.
Thanks for the PR!

**Review by popzxc — APPROVED** [review]

**Pull request merged**